asset/src/_kafka_clients/producer-client.ts

```diff
@@ -63,16 +65,28 @@ export default class ProducerClient extends BaseClient<kafka.Producer> {
 
         const total = messages.length;
         const endOfBatchIndex = (total - 1);
-        const endofBufferIndex = (this._bufferSize - 1);
-        this._logger.debug(`producing ${total} messages in batches ${Math.floor(total / this._bufferSize)}...`);
+        const endofBufferIndex = (this._maxBufferMsgLength - 1);
+        // This is a counter that will track the bytes written in the current batch
+        let currentBatchSizeInBytes = 0;
+        const maxQueueByteSize = this._maxBufferKilobyteSize * 1024;
+        this._logger.debug(`producing ${total} messages in batches ${Math.floor(total / this._maxBufferMsgLength)}...`);
 
         try {
             // Send the messages, after each buffer size is complete
             // flush the messages
             for (let i = 0; i < total; i++) {
                 const msg = messages[i];
                 const message: ProduceMessage = (map == null) ? msg : map(msg);
-                this._bytesProduced += Buffer.byteLength(message.data);
+                const messageByteSize = Buffer.byteLength(message.data);
+                this._bytesProduced += messageByteSize;
+                currentBatchSizeInBytes += messageByteSize;
+
+                // If the current queue batch size gets full, do a flush.
+                if (currentBatchSizeInBytes >= maxQueueByteSize) {
+                    this._logger.warn(`Producer queue is full! Flushing queue before writing more messages...`);
```

Comment:
This is too wishy washy, "queue is full" is ambiguous, is it too many records or two many bytes?  In fact, you have a great opportunity here to share some exact details.  You can tell us `currentBatchSizeInBytes` and `maxQueueByteSize`.  Maybe even convert it back to the original setting and remind us the name of `_maxBufferKilobyteSize` ... the thing it comes from in the job config.